### ISSUE TITLE
Specifying edge index / adjacency

### ISSUE THREAD

**Issue by anton-bushuiev** [lucidrains/equiformer-pytorch#10]
Hi, @lucidrains! Thank you for your impressive work!

Equiformer, as seen in the architecture figure, takes a graph as an input. I am, however, not sure how to specify custom edges (adjacency matrix) to `Equiformer.forward`. I see that it is already implemented but I am still confused how I should properly use `edges` and `neighbor_mask`. I would be very grateful if you could add a simple example with a minimalistic graph.

**Comment by lucidrains** [comment]
Hi Anton! Yup, I can being in edges support

Which of the features under https://github.com/lucidrains/se3-transformer-pytorch#edges most aligns with your usecase, just so I can prioritize a bit

**Comment by anton-bushuiev** [comment]
Thank you for your quick reply! In my case, I have additionally a predefined adjacency given by `edge_index` in COO format from torch-geometric (the edges are quite sparse). So, it aligns with [`Sparse Neighbors`](https://github.com/lucidrains/se3-transformer-pytorch#sparse-neighbors) the best :).

**Comment by lucidrains** [comment]
@anton-bushuiev ok, i'll take care of it, if not by Friday, by weekend's end

i'm quite excited about this so(3) -> so(2) paper

**Comment by lucidrains** [comment]
hey, sorry for the delay, will get around to this either tonight or tomorrow morning

**Comment by anton-bushuiev** [comment]
Sure, thank you! I am currently experimenting with the default k-NN graph setup. I found some minor bug, so I'll make a PR.

**Comment by lucidrains** [comment]
@anton-bushuiev  😦 oh my, i hope it is minor since a reputable lab is relying on the code here.. do submit a PR!

got the edges and adjacency matrix functionality ported over from se3 transformers this morning! https://github.com/lucidrains/equiformer-pytorch#edges let me know if it works for your use case

**Comment by anton-bushuiev** [comment]
Hi, @lucidrains !

Yes, it is indeed minor 🙂. I have opened a [PR](https://github.com/lucidrains/equiformer-pytorch/pull/11).

Thank you so much for adding the support for adjacency! It seems to work well but I cannot figure out why I get all-zero type-1 output if I run the adjacency matrix example. Is it expected? If I set `num_neighbors > 1` it returns random noise as I would expect.
![image](https://github.com/lucidrains/equiformer-pytorch/assets/67932762/6efbf95e-5d4b-4b4f-acca-e65c35e6680c)

**Comment by lucidrains** [comment]
@anton-bushuiev 🤦  my bad, do you want to try again?

**Comment by anton-bushuiev** [comment]
Hi @lucidrains, thanks a lot for fixing! Now, I am able to achieve stable and reasonable training with multiple hyperparameter setups for my problem on protein structures.

I however suspect that there are currently some major issues with padding that occur even without using sparse neighbors:
- Padding tokens influence k-NN graph construction (example 1 in [the notebook on my debug branch](https://github.com/anton-bushuiev/equiformer-pytorch/blob/debug/debug.ipynb))
- Coordinates of padded tokens seem to be involved in gradient computation (examples 2 and 3 in the notebook). Even when coordinates of padded tokens are set to zero instead of inf, [this eps](https://github.com/lucidrains/equiformer-pytorch/commit/295139f6bc28dab7879e1ff7efb3035da92540d0) is required to achieve stable training.

I will try to investigate it deeper.

**Comment by lucidrains** [comment]
@anton-bushuiev thank you for the feedback

believe there was indeed an issue where neighbor masking was not applied within `MLPAttention`

the stability you see with the extra epsilon lines up with the observations of [another paper out of Waymo research](https://github.com/lucidrains/VN-transformer)

**Comment by anton-bushuiev** [comment]
Perfect, thank you a lot @lucidrains!

I am not familiar with VN-Transformer but yes, you are right, with 0.3.9 the eps is not required anymore. So, the instability was caused by padding issues rather than some kind of exact equivariance.

**Comment by lucidrains** [comment]
hey that's great! please email me if you see any interesting results (I promise I won't tell anyone, so you won't get scooped etc)

just so I can know whether to sink more time into this research direction

**Comment by lucidrains** [comment]
regardless, thank you for this issue. really ironed out the remaining kinks
